p2panda-rs/src/graph/graph.rs

```diff
@@ -0,0 +1,495 @@
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+use super::GraphError;
+
+/// Directed acyclic casaul graph which can be sorted topologically.
+///
+/// Graph API based on [tangle-graph](https://gitlab.com/tangle-js/tangle-graph).
+#[derive(Debug, PartialEq, Clone)]
+pub struct Graph<T: PartialEq + Clone + Debug>(HashMap<String, Node<T>>);
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Node<T: PartialEq + Clone + Debug> {
+    key: String,
+    data: T,
+    previous: Vec<String>,
+    next: Vec<String>,
+}
+
+#[derive(Debug, PartialEq, Clone, Default)]
+pub struct GraphData<T: PartialEq + Clone + Debug> {
+    sorted: Vec<T>,
+    merged_branch_tips: Vec<T>,
+    graph_tips: Vec<T>,
+}
+
+impl<T: PartialEq + Clone + Debug> GraphData<T> {
+    // Returns the data from sorted graph nodes.
+    pub fn sorted(&self) -> Vec<T> {
+        self.sorted.clone()
+    }
+    // Returns the current tips of this graph.
+    pub fn current_graph_tips(&self) -> Vec<T> {
+        self.graph_tips.clone()
+    }
+    // Returns a list containing all branch tips and the current graph tips.
+    pub fn all_graph_tips(&self) -> Vec<T> {
+        let mut all_graph_tips = self.graph_tips.clone();
+        all_graph_tips.extend(self.merged_branch_tips.clone());
+        all_graph_tips
+    }
+}
+
+impl<'a, T: PartialEq + Clone + Debug> Node<T> {
+    /// Returns true if this node is the root of this graph.
+    fn is_root(&self) -> bool {
+        self.previous.is_empty()
+    }
+
+    /// Returns true if this is a merge node.
+    fn is_merge(&self) -> bool {
+        self.previous.len() > 1
+    }
+
+    /// Returns true if this is a branch node.
+    fn is_branch(&self) -> bool {
+        self.next.len() > 1
+    }
+
+    /// Returns true if this is a graph tip.
+    fn is_tip(&self) -> bool {
+        self.next.is_empty()
+    }
+
+    /// Returns the key for this node.
+    fn key(&self) -> String {
+        self.key.to_owned()
+    }
+
+    /// Returns a vector of keys for the nodes preceding this node in the graph.
+    fn previous(&self) -> Vec<String> {
```

Comment:
General Rust thing I was wondering about: wouldn't it be better to generally return references instead of cloning for such getter methods because then we don't need to make a copy in memory whenever we want to read that value?

Reply:
Yes, I missed these, thanks!

Reply:
Generally agreeing with your comment, in that case though the contents of a `Vec` need to be owned as far as I know. One could return a reference to the whole `Vec` but that brings ownership problems later on with iterators.